### ISSUE TITLE
Verilog grammar current design are unable to allow us for doing a preprocessor for verilator

### ISSUE THREAD

**Issue by quantrpeter** [antlr/grammars-v4#2663]
Hi all
    Verilog grammar current design is unable to allow us for doing a preprocessor for verilator. Here is what I want to do:

1. I want to build a preprocess to insert some code to the source code and then pass it to verilator for compilation
2. Existing grammar need us to preprocess all macro, replacing all macro with the value before we can use antlr to parse it. So it comes with a problem, our output files still remind having macro and let verilator to process the macro, rather than we replace all macro.

Anyway for work around?

thanks
Peter

**Comment by msagca** [comment]
Hi @quantrpeter,

I don't quite understand the problem here. Preprocessing is something you have to take care of, ANTLR won't do it for you.

You need to tokenize the input and all `include` files (up to a maximum nesting level) first. Then, pass those token streams to `VerilogPreParser` and the resulting parse trees to a user application (that extends the base listener or visitor class) to do preprocessing. If the inputs contain macro calls, you have to replace those with the associated `macro_text` (using [TokenStreamRewriter](https://www.antlr.org/api/Java/org/antlr/v4/runtime/TokenStreamRewriter.html)). If there are conditional compilation directives (`ifdef`, `else`, etc.) you have to resolve them and make the necessary changes to the original input text. After these (text) replacements, you can send the preprocessed text to `Verilator` or tokenize it again and pass the stream to `VerilogParser`.

I'm sure there are better ways to do it but this is how I do preprocessing. I tried to explain the process in more detail [here](https://github.com/msagca/grammars-v4/issues/2).

**Comment by quantrpeter** [comment]
the complexity goes high if the macro is a function and its variables/parameters are named so close to each other. The string template method you mentioned in the above link won't work.

Why don't we create a grammar to see all macro? it won't be much easier for people to build verilog pre-processor.

**Comment by msagca** [comment]
@quantrpeter,

Writing better macro call and macro definition rules (so people can easily access each macro argument) requires me to embed actions into the grammar; however, in that case, the grammar would be target language dependent, which is against the guidelines/philosophy of this repository.

There are ANTLR4 grammars on GitHub that satisfy your requirements such as [these](https://github.com/Nic30/hdlConvertor/tree/master/grammars). Maybe you should give them a try.

Also, can you give an example macro call where you can't extract the argument information from the text using regex or some other methods in the target language?

**Comment by quantrpeter** [comment]
Hi, i got this macro

```
    `define MEMWB_SETTINGS(we, wa, wdata, cycle) \
        wb_wreg     <= we; \
        wb_wd       <= wa; \
        wb_wdata    <= wdata; \
        stall_cycle <= cycle;
```

I can extract parameter by regex, i i know **we, wa, wdata, cycle**, now i need to replace the value to the macro body, if I call the macro like this:

```
`MEMWB_SETTINGS(`Disable,`ZeroRegAddr,`ZeroDWord,`ZeroCycle)
```

`Disable will become we and `ZeroRegAddr will become wa. Since the macro parameter can be in LHS or RHS, and we don;t have a grammar to parse the macro body, if just replace it by string, it is danger. thanks

**Comment by msagca** [comment]
It is impossible to come up with a grammar that can parse macro body because it's just text, it doesn't follow any language rules. Macro calls can be placed anywhere in the input. We can only tokenize and parse it after replacing all the calls with the text.

**Comment by quantrpeter** [comment]
@msagca agree

**Comment by kaby76** [comment]
Currently, preprocessor problems aren't handled in this repo. E.g., csharp/, cobol85/, objc/ and, of course, verilog/. But, it's not because it can't be handled by the current framework (not quite true because of issues with Antlr and the Go target), or that it goes against action-free grammars. We have plenty of grammars in the repo that have actions. But, these grammars are in "target agnostic format", and they only evaluate the semantics in so far as influencing the parse.

It's possible to write a grammar with a rule added that implements a preprocessor step in an action, then calls the entry point for the grammar that is post-processed text. The main problem is that it's hard to write a good preprocessor. But, I've been doing this for a replacement of the cpp/ grammar.

**Comment by quantrpeter** [comment]
@msagca if merge text_macro_definition and text_macro_usage back to VerilogParser.g4, will it work?

**Comment by msagca** [comment]
@quantrpeter 

You cannot merge those into VerilogParser grammar because those structures may appear in weird places, they are not a part of Verilog source text.

**Comment by quantrpeter** [comment]
after two weeks fighting, I finally make it work. project in https://gitlab.com/quantr/toolchain/verilog-compiler
Verilog preprocess function here https://gitlab.com/quantr/toolchain/verilog-compiler/-/blob/main/src/main/java/hk/quantr/verilogcompiler/VerilogHelper.java

thanks
